storage/postgres/log_storage.go

```diff
@@ -0,0 +1,999 @@
+	}
+	if len(results) != len(toRetrieve) {
+		return nil, fmt.Errorf("failed to retrieve all existing leaves: got %d, want %d", len(results), len(toRetrieve))
+	}
+	// Replace the requested leaves with the actual leaves.
+	for i, requested := range existingLeaves {
+		if requested == nil {
+			continue
+		}
+		found := false
+		for _, result := range results {
+			if bytes.Equal(result.LeafIdentityHash, requested.LeafIdentityHash) {
+				existingLeaves[i] = result
+				found = true
+				break
+			}
+		}
+		if !found {
+			return nil, fmt.Errorf("failed to find existing leaf for hash %x", requested.LeafIdentityHash)
+		}
+	}
+	totalDuration := time.Since(start)
+	readDuration := totalDuration - insertDuration
+	observe(queueReadLatency, readDuration, label)
+	observe(queueLatency, totalDuration, label)
+
+	return existingLeaves, nil
+}
+
+func (t *logTreeTX) AddSequencedLeaves(ctx context.Context, leaves []*trillian.LogLeaf, timestamp time.Time) ([]*trillian.QueuedLogLeaf, error) {
+	res := make([]*trillian.QueuedLogLeaf, len(leaves))
+	ok := status.New(codes.OK, "OK").Proto()
+
+	// Leaves in this transaction are inserted in two tables. For each leaf, if
+	// one of the two inserts fails, we remove the side effect by rolling back to
+	// a savepoint installed before the first insert of the two.
+	const savepoint = "SAVEPOINT AddSequencedLeaves"
+	if _, err := t.tx.ExecContext(ctx, savepoint); err != nil {
+		glog.Errorf("Error adding savepoint: %s", err)
+		return nil, err
+	}
+	// TODO(pavelkalinnikov): Consider performance implication of executing this
+	// extra SAVEPOINT, especially for 1-entry batches. Optimize if necessary.
+
+	// Note: LeafData inserts are presumably protected from deadlocks due to
+	// sorting, but the order of the corresponding SequencedLeafData inserts
+	// becomes indeterministic. However, in a typical case when leaves are
+	// supplied in contiguous non-intersecting batches, the chance of having
+	// circular dependencies between transactions is significantly lower.
+	ordLeaves := sortLeavesForInsert(leaves)
+	for _, ol := range ordLeaves {
+		i, leaf := ol.idx, ol.leaf
+
+		// This should fail on insert, but catch it early.
+		if got, want := len(leaf.LeafIdentityHash), t.hashSizeBytes; got != want {
+			return nil, status.Errorf(codes.FailedPrecondition, "leaves[%d] has incorrect hash size %d, want %d", i, got, want)
+		}
+
+		if _, err := t.tx.ExecContext(ctx, savepoint); err != nil {
+			glog.Errorf("Error updating savepoint: %s", err)
+			return nil, err
+		}
+
+		res[i] = &trillian.QueuedLogLeaf{Status: ok}
+
+		// TODO(pavelkalinnikov): Measure latencies.
+		_, err := t.tx.ExecContext(ctx, insertLeafDataSQL,
+			t.treeID, leaf.LeafIdentityHash, leaf.LeafValue, leaf.ExtraData, timestamp.UnixNano())
+		// TODO(pavelkalinnikov): Detach PREORDERED_LOG integration latency metric.
+
+		// TODO(pavelkalinnikov): Support opting out from duplicates detection.
+		//if isDuplicateErr(err) {
```

Comment:
Bunch of commented-out code here, remove?

Reply:
Can do